src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/StandardScaler.java

```diff
@@ -0,0 +1,64 @@
+package org.mastodon.mamut.feature.dimensionalityreduction.umap.util;
+
+import org.apache.commons.math3.stat.StatUtils;
+
+/**
+ * Standardize vectors by removing the mean and scaling to unit variance.
+ * <br>
+ * Re-implemented from scikit-learn's <a href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html">StandardScaler</a>.
+ * <br>
+ * Cf. as well <a href="https://scikit-learn.org/stable/modules/preprocessing.html">Preprocessing data</a>.
+ */
+public class StandardScaler
+{
+	private StandardScaler()
+	{
+		// prevent instantiation
+	}
+
+	/**
+	 * Standardizes a specific column of the given 2D array (matrix).
+	 * <br>
+	 * This method extracts the specified column from the input matrix, standardizes it by removing the mean and scaling to unit variance,
+	 * and then replaces the original column in the matrix with the standardized values.
+	 *
+	 * @param matrix The 2D array whose column is to be standardized.
+	 * @param columnIndex The index of the column to standardize.
+	 */
+	private static void standardizeColumn( double[][] matrix, int columnIndex )
+	{
+		int numRows = matrix.length;
+		double[] column = new double[ numRows ];
+
+		for ( int i = 0; i < numRows; i++ )
+		{
+			column[ i ] = matrix[ i ][ columnIndex ];
+		}
+		double[] standardizedColumn = StatUtils.normalize( column );
+
+		for ( int i = 0; i < numRows; i++ )
+		{
+			matrix[ i ][ columnIndex ] = Double.isNaN( standardizedColumn[ i ] ) ? 0d : standardizedColumn[ i ]; // replace NaNs with 0
```

Comment:
This line removes any NaN value in the matrix. I guess it is implemented like this on the assumption, that all NaN values originate from StatUtils.normalize when applied on a constant column. It is ok to make this assumption, but only if you assume that the input matrix has NaN. Which is also true in this code base. But please document, that the standardizeColumns method expects the input matrix/array to have no NaN values.

Reply:
This line could benefit from a comment, like:
_StatUtils.normalize(...) will write NaN if a column has variance=0, it is important to replace this NaN values before computing the UMAP_